packages/core/src/utils/userop.ts

```diff
@@ -1,32 +1,104 @@
+import type { Address } from "abitype";
+import {
+  concat,
+  encodeAbiParameters,
+  hexToBigInt,
+  isAddress,
+  keccak256,
+  pad,
+  toHex,
+  type Hash,
+  type Hex,
+} from "viem";
+import type { EntryPointVersion } from "../entrypoint/types";
+import { UnexpectedEntryPointVersionError } from "../errors/entrypoint.js";
 import type {
   BigNumberish,
   Multiplier,
   UserOperationFeeOptionsField,
   UserOperationRequest,
+  UserOperationRequest_v7,
   UserOperationStruct,
+  UserOperationStruct_v6,
+  UserOperationStruct_v7,
 } from "../types";
 import { bigIntClamp, bigIntMultiply } from "./bigint.js";
-import { isBigNumberish } from "./index.js";
+import { allEqual, isBigNumberish } from "./index.js";
 
 /**
- * Utility method for asserting a {@link UserOperationStruct} is a {@link UserOperationRequest}
+ * Utility method for asserting a {@link UserOperationStruct} has valid fields for the given entry point version
  *
  * @param request a {@link UserOperationStruct} to validate
- * @returns a type guard that asserts the {@link UserOperationStruct} is a {@link UserOperationRequest}
+ * @param entryPointVersion a {@link EntryPointVersion} entry point version
+ * @returns a type guard that asserts the {@link UserOperationRequest} is valid
  */
-export function isValidRequest(
-  request: UserOperationStruct
-): request is UserOperationRequest {
+export function isValidRequest<
+  TEntryPointVersion extends EntryPointVersion = EntryPointVersion
+>(
+  request: UserOperationStruct<TEntryPointVersion>
+): request is UserOperationRequest<TEntryPointVersion> {
   // These are the only ones marked as optional in the interface above
   return (
-    !!request.callGasLimit &&
-    !!request.maxFeePerGas &&
-    request.maxPriorityFeePerGas != null &&
-    !!request.preVerificationGas &&
-    !!request.verificationGasLimit
+    BigInt(request.callGasLimit || 0n) > 0n &&
+    BigInt(request.maxFeePerGas || 0n) > 0n &&
+    BigInt(request.maxPriorityFeePerGas || 0n) > 0n &&
+    BigInt(request.preVerificationGas || 0n) > 0n &&
+    BigInt(request.verificationGasLimit || 0n) > 0n &&
+    isValidPaymasterAndData(request) &&
+    isValidFactoryAndData(request)
   );
 }
 
+/**
+ * Utility method for asserting a {@link UserOperationRequest} has valid fields for the paymaster data
+ *
+ * @param request a {@link UserOperationRequest} to validate
+ * @param entryPointVersion a {@link EntryPointVersion} entry point version
+ * @returns a type guard that asserts the {@link UserOperationRequest} is a {@link UserOperationRequest}
+ */
+export function isValidPaymasterAndData<
+  TEntryPointVersion extends EntryPointVersion = EntryPointVersion
+>(request: UserOperationStruct<TEntryPointVersion>): boolean {
+  if (!("paymaster" in request)) {
+    const { paymasterAndData } = request as UserOperationStruct_v6;
+    return paymasterAndData != null;
```

Comment:
for this paymasterAndData must be set as `0x` if it's not being used. Tested this out with @adam-alchemy the other day and we got an error from the RPC